knora/dsplib/utils/stack_handling.py

```diff
@@ -0,0 +1,122 @@
+import json
+import re
+import subprocess
+import time
+from pathlib import Path
+from typing import Optional
+
+import requests
+
+from knora.dsplib.models.helpers import BaseError
+
+# relative path to "knora/dsplib/docker", to make it accessible when dsp-tools is called from another working directory
+docker_path = Path(__file__).parent / Path("../docker")
+
+
+def start_stack(
+    max_file_size: Optional[int] = None,
+    enforce_docker_system_prune: bool = False,
+    suppress_docker_system_prune: bool = False
+) -> None:
+    """
+    Start the Docker containers of DSP-API and DSP-APP, and load some basic data models and data. After startup, ask
+    user if Docker should be pruned or not.
+
+    Args:
+        max_file_size: max. multimedia file size allowed by SIPI, in MB (max: 100'000)
+        enforce_docker_system_prune: if True, prune Docker without asking the user
+        suppress_docker_system_prune: if True, don't prune Docker (and don't ask)
+    """
+    if enforce_docker_system_prune and suppress_docker_system_prune:
+        raise BaseError('The arguments "--prune" and "--no-prune" are mutually exclusive')
+
+    # get sipi.docker-config.lua
+    latest_release = json.loads(requests.get("https://api.github.com/repos/dasch-swiss/dsp-api/releases").text)[0]
+    url_prefix = f"https://github.com/dasch-swiss/dsp-api/raw/{latest_release['target_commitish']}/"
+    docker_config_lua_text = requests.get(f"{url_prefix}sipi/config/sipi.docker-config.lua").text
+    if max_file_size is not None:
+        if not 1 <= max_file_size <= 100_000:
+            raise BaseError("max_file_size must be between 1 and 100000")
+        max_post_size_regex = r"max_post_size ?= ?[\'\"]\d+M[\'\"]"
+        if not re.search(max_post_size_regex, docker_config_lua_text):
+            raise BaseError("sipi.docker-config.lua doesn't contain a variable max_post_size that could be replaced")
```

Comment:
Thanks for the hints. One of my long-term ideas is to raise different kinds of exceptions in the code, and then catching the user-feedback-exceptions in `dsp_tools.py`, print the message, and then `exit()` without stack trace. Only the "real errors" would escalate uncaught and lead to a stack trace. Do you think that would be a good design pattern?

Or even easier: if an error occurs which is the user's fault, print message and `exit()`. Use BaseError only if an internal-caused error occurs.

I will do that in a separate PR, but what do you think about it?

Reply:
That's definitely a reasonable and common approach. You can then create an inheritance hierarchy of exceptions for different categories of errors, and things like that.  
And it will definitely be a great improvement to how it is currently.

Personally I'm not the biggest fan of exceptions, but in Python it's not exactly easy to find better solutions. To me, exceptions always feel like "blowing up stuff", and there is nothing very "exceptional" about a user entering a wrong parameter... But as I said, there is no super-neat pythonic way of a controlled and typed error channel like we have in ZIO.